### PR TITLE
Add kpod import command

### PR DESCRIPTION
**Pull request by umohnani8**
Imports a tarball and saves it as filesystem image.
One error message change to kpod history
Vendor in latest containers/image

Signed-off-by: Urvashi Mohnani <umohnani@redhat.com>

### CODE REVIEW

**Comment by umohnani8** [comment]
@rhatdan PTAL

File: cmd/kpod/import.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	}
+	importDescription = "Imports a tarball and saves it as a root filesystem image.\n" +
+		"The commit message and image config can be modified by the user."
+	importCommand = cli.Command{
+		Name:        "import",
+		Usage:       "Import a tarball to create a filesystem image",
+		Description: importDescription,
+		Flags:       importFlags,
+		Action:      importCmd,
+		ArgsUsage:   "TARBALL",
```
Comment (rhatdan) [inline cmd/kpod/import.go]: Show me the optional reference.

[ reference ]

File: cmd/kpod/import.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+		cli.StringFlag{
+			Name:  "change, c",
+			Usage: "Apply imgspecv1 configurations to the created image",
+		},
+		cli.StringFlag{
+			Name:  "message, m",
+			Usage: "Set commit message for the image imported",
+		},
+	}
+	importDescription = "Imports a tarball and saves it as a root filesystem image.\n" +
```
Comment (rhatdan) [inline cmd/kpod/import.go]: Use the ` 

`
Syntax.

File: cmd/kpod/import.go [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+		source = args[0]
+		opts.Reference = args[1]
+	default:
+		return errors.Errorf("too many arguments, need 2 only")
+	}
+
+	if _, err := url.ParseRequestURI(source); err == nil {
+		file, err := downloadFromURL(source)
+		if err != nil {
+			return nil
```
Comment (rhatdan) [inline cmd/kpod/import.go]: Return nil or err?

File: cmd/kpod/import.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+	case 1:
+		source = args[0]
+	case 2:
+		source = args[0]
+		opts.Reference = args[1]
+	default:
+		return errors.Errorf("too many arguments, need 2 only")
+	}
+
+	if _, err := url.ParseRequestURI(source); err == nil {
```
Comment (rhatdan) [inline cmd/kpod/import.go]: What happens if err != nil?

File: libpod/runtime_img.go [hunk 5]
```diff
@@ -877,8 +885,70 @@ func (r *Runtime) GetHistory(image string) ([]ociv1.History, []types.BlobInfo, s
...
+	}
+
+	var reference = options.Reference
+
+	sc := common.GetSystemContext("", "")
+
+	if reference == "" {
+		newImg, err := src.NewImage(sc)
+		if err != nil {
+			return err
```
Comment (rhatdan) [inline libpod/runtime_img.go]: Should be wrapped?

File: libpod/runtime_img.go [hunk 6]
```diff
@@ -877,8 +885,70 @@ func (r *Runtime) GetHistory(image string) ([]ociv1.History, []types.BlobInfo, s
...
+		digest := newImg.ConfigInfo().Digest
+		if err = digest.Validate(); err != nil {
+			return errors.Wrapf(err, "error getting config info")
+		}
+		reference = "@" + digest.Hex()
+	}
+
+	policy, err := signature.DefaultPolicy(sc)
+	if err != nil {
+		return err
```
Comment (rhatdan) [inline libpod/runtime_img.go]: Should this be wrapped?

File: libpod/runtime_img.go [hunk 7]
```diff
@@ -877,8 +885,70 @@ func (r *Runtime) GetHistory(image string) ([]ociv1.History, []types.BlobInfo, s
...
+	}
+
+	policyContext, err := signature.NewPolicyContext(policy)
+	if err != nil {
+		return err
```
Comment (rhatdan) [inline libpod/runtime_img.go]: Wrapped?

**Comment by mheon** [comment]
I think you need to redo vendoring the update c/image. A lot more than just containers/image was updated.

File: cmd/kpod/import.go (hunk 2 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: fixed

File: cmd/kpod/import.go (hunk 4 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: fixed

File: cmd/kpod/import.go (hunk 3 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: should be err, fixed

File: libpod/runtime_img.go (hunk 5 above)
Comment (umohnani8) [reply libpod/runtime_img.go]: wrapped.

File: libpod/runtime_img.go (hunk 6 above)
Comment (umohnani8) [reply libpod/runtime_img.go]: wrapped

File: libpod/runtime_img.go (hunk 7 above)
Comment (umohnani8) [reply libpod/runtime_img.go]: wrapped

**Comment by umohnani8** [comment]
@mheon I just did containers/image because I needed the new tarball transport that was added. What else should I update in the vendor?

**Comment by mheon** [comment]
@umohnani8 The problem is that you picked up too much, not too little - the kubernetes and CNI vendors also updated. I think you need to specifically update only the containers/image vendor with vndr (specify the repo and updated commit directly to vndr). Alternatively, you could directly revert the changes to all the vendors except containers/image by directly manipulating the vendor folder.

**Comment by umohnani8** [comment]
@mheon thant is what I did. I changed the commit id of containers/image to match the tarball commit and just vendored that it. I can try it again.

**Comment by mheon** [comment]
@umohnani8 A few of the other vendors (kubernetes, CNI to name a few) don't have specific commits IDs attached, so they have a tendency to update whenever we vendor something new in. I think @runcom and @mrunalp are familiar with how to work around this - might want to ask them for help?

**Comment by umohnani8** [comment]
So when I do `vndr github.com/containers/image` with the tarball commit id, it shows me a bunch of files have been deleted, which include the Kubernetes and CNI files.
I just selected the containers/image files and committed those.

**Comment by mheon** [comment]
Vendoring looks fine now, thanks @umohnani8 

I'll do a more thorough review after lunch

**Comment by umohnani8** [comment]
thanks for pointing that out @mheon :)

**Comment by mheon** [comment]
Bunch of pull-related tests seem broken. Unclear if this is the new CI acting up or an actual regression.

**Comment by umohnani8** [comment]
@mheon the tests should pass now. Something with the new containers/image was causing them to break.
I am not sure what exactly is happening with containers/image right now, but I just vendored in the changes for the tarball transport to support kpod import. @mrunalp @runcom is that acceptable?

**Comment by rhatdan** [comment]
Well I like the fact that you got rid of Nalins pool

**Comment by mheon** [comment]
@umohnani8 Containers/image broke a lot of kpod image-related functions, so you dropped the parts that broke us? I'm a bit concerned that we're broken on the upstream code - we should try and figure out how to fix either c/image or how we use c/image so we can use all the updated code.

**Comment by umohnani8** [comment]
@mheon I don't believe I dropped anything that we already had. I just added the tarball folder and changes to the transports folder. There were other changes to containers/image, which was breaking our tests so I didn't commit those after vendoring in containers/image.

**Comment by mheon** [comment]
@umohnani8 I'm just a little iffy on cherry picking a partial new commit from c/image. I think we should bother someone on the image side to ask why the new commit broke kpod.

**Comment by umohnani8** [comment]
@mheon the main thing that was breaking was pulling a docker-archive file - it kept throwing the error "blob size mismatch". There was a PR in containers/image that got merged when it shouldn't have and I spoke to @nalind, but he said most of the changes to that was for containers-storage, so weird that docker-archive is messing up.
I can pull in the changes and try to get to the bottom of why its causing pulling from docker-archive to fail.

**Comment by mheon** [comment]
@umohnani8 If it's not too much trouble, sure - I'd just like to make sure we aren't going to hit this again next time we have to update c/image.

**Comment by rhatdan** [comment]
I agree with Matt, I would rather get this cleaned up in containers/image then get kpod import in.

**Comment by umohnani8** [comment]
All tests should pass now. I was vendoring in from containers/image master and not @nalind's fork, which was causing compression issues with docker-archive therefore breaking the tests.
Vendored in latest from Nalin's fork and all tests should pass now.
@mheon @rhatdan PTAL.

File: cmd/kpod/import.go [hunk 8]
```diff
@@ -0,0 +1,192 @@
...
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/projectatomic/libpod/libpod"
+
```
Comment (rhatdan) [inline cmd/kpod/import.go]: Remove space

File: cmd/kpod/import.go [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+	"github.com/opencontainers/image-spec/specs-go/v1"
+
```
Comment (rhatdan) [inline cmd/kpod/import.go]: remove space

File: cmd/kpod/import.go [hunk 10]
```diff
@@ -0,0 +1,192 @@
...
+	"github.com/pkg/errors"
+	"github.com/urfave/cli"
+)
+
+var (
+	importFlags = []cli.Flag{
+		cli.StringFlag{
+			Name:  "change, c",
+			Usage: "Apply imgspecv1 configurations to the created image",
```
Comment (rhatdan) [inline cmd/kpod/import.go]: Docker says:
  -c, --change list      Apply Dockerfile instruction to the created image (default [])

I have no idea what a imgspecv1 configurations are.


Man page says
```
       -c, --change=[]
          Apply specified Dockerfile instructions while importing the image
          Supported Dockerfile instructions: CMD|ENTRYPOINT|ENV|EXPOSE|ONBUILD|USER|VOLUME|WORKDIR
```

File: cmd/kpod/import.go [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+		},
+		cli.StringFlag{
+			Name:  "message, m",
+			Usage: "Set commit message for the image imported",
```
Comment (rhatdan) [inline cmd/kpod/import.go]: `Set commit message for imported image`
Reads better

File: cmd/kpod/import.go [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+		},
+	}
+	importDescription = `Imports a tarball and saves it as a root filesystem image.
```
Comment (rhatdan) [inline cmd/kpod/import.go]: Create a container image from the contents of the specified tarball  (.tar, .tar.gz, .tgz, .bzip, .tar.xz, .txz).  
Note remote tar balls can be specified, via web address.

Optionally tag the image.  You can specify the Dockerfile instructions using the --change option.

File: cmd/kpod/import.go [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+		Usage:       "Import a tarball to create a filesystem image",
+		Description: importDescription,
+		Flags:       importFlags,
+		Action:      importCmd,
+		ArgsUsage:   "TARBALL [REFERENCE]",
+	}
+)
+
+func importCmd(c *cli.Context) error {
+	if err := validateFlags(c, historyFlags); err != nil {
```
Comment (rhatdan) [inline cmd/kpod/import.go]: historyFlags?

File: cmd/kpod/import.go [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+		return errors.Wrapf(err, "Could not get runtime")
+	}
+	defer runtime.Shutdown(false)
+
+	var opts libpod.CopyOptions
+	var source string
+	args := c.Args()
+	switch len(args) {
+	case 0:
+		return errors.Errorf("need to give the path to the tarball")
```
Comment (rhatdan) [inline cmd/kpod/import.go]: Can you specify "-"

"must specify a tarball of '-' for stdin'

File: cmd/kpod/import.go [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+	case 1:
+		source = args[0]
+	case 2:
+		source = args[0]
+		opts.Reference = args[1]
+	default:
+		return errors.Errorf("too many arguments, need 2 only")
```
Comment (rhatdan) [inline cmd/kpod/import.go]: too many arguments.  Usage tarball [reference]

File: cmd/kpod/import.go [hunk 16]
```diff
@@ -0,0 +1,192 @@
...
+	}
+
+	// if source is a url, download it and save to a temp file
+	u, err := url.ParseRequestURI(source)
+	if err == nil && u.Scheme != "" {
+		file, err := downloadFromURL(source)
+		if err != nil {
+			return err
```
Comment (rhatdan) [inline cmd/kpod/import.go]: This should be done after all other parsing, so we don't pull an image and then fail later.

File: cmd/kpod/import.go (hunk 10 above)
Comment (rhatdan) [reply cmd/kpod/import.go]: Change should be a stringsliceflag.  You can specify it multiple times.

File: cmd/kpod/import.go [hunk 17]
```diff
@@ -0,0 +1,192 @@
...
+		}
+		defer os.Remove(file)
+		source = file
+	}
+
+	changes := v1.ImageConfig{}
+	if c.IsSet("change") {
```
Comment (rhatdan) [inline cmd/kpod/import.go]: Change is a slice, should be a for loop.

File: docs/kpod-import.1.md [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+**kpod [GLOBAL OPTIONS]**
+
+**kpod import [GLOBAL OPTIONS]**
+
+**kpod import [OPTIONS] CONTAINER**
+
+## OPTIONS
+
+**--change, -c**
+Apply imgspecv1 configurations to the created image
```
Comment (rhatdan) [inline docs/kpod-import.1.md]: Dockerfile configurations.

Comment (rhatdan) [reply docs/kpod-import.1.md]: Can be specified multiple times.

File: libpod/runtime_img.go [hunk 19]
```diff
@@ -1080,3 +1121,32 @@ func findImageInSlice(images []storage.Image, ref string) (storage.Image, error)
 	}
 	return storage.Image{}, errors.New("could not find image")
 }
+
+// getIMgeDigest creates an image object and uses the hex value of the digest as the image ID
```
Comment (rhatdan) [inline libpod/runtime_img.go]: Spelling.

File: cmd/kpod/import.go (hunk 10 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: Yeah, I wasn't sure if we wanted to call it docker file instructions. The struct that holds all these configurations is the imgspecv1.Image struct so I used that.
Should I rephrase it to be dockerfile instructions or?

Comment (rhatdan) [reply cmd/kpod/import.go]: Well instead of saying Dockerfile, why not list the possible options.

File: cmd/kpod/import.go (hunk 8 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: fixed

File: cmd/kpod/import.go (hunk 9 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: fixed

File: cmd/kpod/import.go (hunk 10 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: fixed

File: cmd/kpod/import.go (hunk 11 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: fixed

File: cmd/kpod/import.go (hunk 12 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: fixed

File: cmd/kpod/import.go (hunk 13 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: woops, fixed.

File: cmd/kpod/import.go (hunk 14 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: fixed

File: cmd/kpod/import.go (hunk 16 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: fixed

File: cmd/kpod/import.go (hunk 17 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: fixed

File: docs/kpod-import.1.md (hunk 18 above)
Comment (umohnani8) [reply docs/kpod-import.1.md]: fixed

File: libpod/runtime_img.go (hunk 19 above)
Comment (umohnani8) [reply libpod/runtime_img.go]: fixed.

File: test/kpod_import.bats [hunk 20]
```diff
@@ -0,0 +1,141 @@
...
+    [ "$status" -eq 0 ]
+    inspect="$output"
+    run grep "/bin/bash" <<< "$inspect"
+    echo "$output"
+    [ "$status" -eq 0 ]
+    cleanup_ctrs
+    cleanup_pods
+    stop_crio
+    rm -f container.tar
+}
```
Comment (rhatdan) [inline test/kpod_import.bats]: newline

**Comment by rhatdan** [comment]
One last nit, LGTM

@mheon @mitrmac PTAL

**Comment by mheon** [comment]
@umohnani8 The code from Nalin's c/image fork is going to land upstream at some point, right? I don't want to be permanently tied to his fork instead of upstream

File: test/kpod_import.bats (hunk 20 above)
Comment (umohnani8) [reply test/kpod_import.bats]: fixed.

File: cmd/kpod/import.go [hunk 21]
```diff
@@ -0,0 +1,190 @@
...
+)
+
+func importCmd(c *cli.Context) error {
+	if err := validateFlags(c, importFlags); err != nil {
+		return err
+	}
+
+	runtime, err := getRuntime(c)
+	if err != nil {
+		return errors.Wrapf(err, "Could not get runtime")
```
Comment (mheon) [inline cmd/kpod/import.go]: Errors should start with a lowercase letter.

@rhatdan Golint should have caught this - are we running it in the tests yet?

**Comment by umohnani8** [comment]
@mheon I believe so. Also Nalin's fork is what has been vendored in for cri-o.

File: cmd/kpod/import.go (hunk 21 above)
Comment (umohnani8) [reply cmd/kpod/import.go]: fixed.

File: cmd/kpod/save.go [hunk 22]
```diff
@@ -91,6 +91,7 @@ func saveCmd(c *cli.Context) error {
 	for _, image := range args {
 		dest := dst + ":" + image
 		if err := runtime.PushImage(image, dest, saveOpts); err != nil {
+			os.Remove(output)
```
Comment (mheon) [inline cmd/kpod/save.go]: This can return an error, should check that

File: libpod/runtime_img.go [hunk 23]
```diff
@@ -877,8 +867,59 @@ func (r *Runtime) GetHistory(image string) ([]ociv1.History, []types.BlobInfo, s
...
+	err = updater.ConfigUpdate(options.ImageConfig, annotations)
+	if err != nil {
+		return errors.Wrapf(err, "error updating image config")
+	}
+
+	var reference = options.Reference
+
+	sc := common.GetSystemContext("", "")
+
+	// 	if reference not given, get the image digest
```
Comment (mheon) [inline libpod/runtime_img.go]: Lot of extra whitespace in this comment

File: libpod/runtime_img.go [hunk 24]
```diff
@@ -877,8 +867,59 @@ func (r *Runtime) GetHistory(image string) ([]ociv1.History, []types.BlobInfo, s
...
+		return errors.Wrapf(err, "error parsing image name %q", path)
+	}
+
+	updater, ok := src.(tarball.ConfigUpdater)
+	if !ok {
+		return errors.Wrapf(err, "unexpected type, a tarball reference should implement tarball.ConfigUpdater")
+	}
+
+	annotations := make(map[string]string)
+	annotations[ociv1.AnnotationDescription] = "test image built"
```
Comment (mheon) [inline libpod/runtime_img.go]: Should this be kept in? Looks like a debugging statement

File: libpod/runtime_img.go [hunk 25]
```diff
@@ -1080,3 +1121,32 @@ func findImageInSlice(images []storage.Image, ref string) (storage.Image, error)
...
+	defer newImg.Close()
+
+	digest := newImg.ConfigInfo().Digest
+	if err = digest.Validate(); err != nil {
+		return "", errors.Wrapf(err, "error getting config info")
+	}
+	return "@" + digest.Hex(), nil
+}
+
+func getPolicyContext(ctx *types.SystemContext) (*signature.PolicyContext, error) {
```
Comment (mheon) [inline libpod/runtime_img.go]: Can we get a brief comment on this function?

File: cmd/kpod/save.go (hunk 22 above)
Comment (umohnani8) [reply cmd/kpod/save.go]: fixed

File: libpod/runtime_img.go (hunk 23 above)
Comment (umohnani8) [reply libpod/runtime_img.go]: fixed

File: libpod/runtime_img.go (hunk 24 above)
Comment (umohnani8) [reply libpod/runtime_img.go]: removed.

File: libpod/runtime_img.go (hunk 25 above)
Comment (umohnani8) [reply libpod/runtime_img.go]: fixed

**Comment by umohnani8** [comment]
@rhatdan tests passed.

File: cmd/kpod/save.go [hunk 26]
```diff
@@ -91,6 +91,9 @@ func saveCmd(c *cli.Context) error {
 	for _, image := range args {
 		dest := dst + ":" + image
 		if err := runtime.PushImage(image, dest, saveOpts); err != nil {
+			if err2 := os.Remove(output); err2 != nil {
```
Comment (rhatdan) [inline cmd/kpod/save.go]: I think you would just logrus this.  But return the original error.

**Comment by rhatdan** [comment]
@mheon PTAL
LGTM other then the comment above.

File: cmd/kpod/save.go (hunk 26 above)
Comment (umohnani8) [reply cmd/kpod/save.go]: fixed

**Comment by mheon** [comment]
LGTM. @umohnani8 Can you add an issue after this merges to switch back to containers/image master once Nalin's changes merge?

**Comment by rhatdan** [comment]
@rh-atomic-bot r+
